### PR TITLE
[SD-2373][SD-2376] ANPA Output Format and file system distribution me…

### PR DESCRIPTION
**Pull request by marwoodandrew**
…thod

### CODE REVIEW

**Comment by sivakuna-aap** [comment]
:+1:

File: server/apps/publish/formatters/anpa_formatter.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            anpa += str(pub_seq_num).zfill(4).encode('ascii')
+            # field seperator
+            anpa += b'\x0A'  # -LF
+            anpa += article.get('priority', 'r').encode('ascii')
+            anpa += b'\x20'
+
+            category = article.get('anpa-category', {})
+            if 'qcode' in category:
+                anpa += article.get('anpa-category', {})['qcode'].encode('ascii')
+            # TODO TODO TODO
```
Comment (mdhaman) [inline server/apps/publish/formatters/anpa_formatter.py]: maybe create a ticket for this TODO.

**Comment by amagdas** [comment]
:+1:

File: server/apps/publish/formatters/anpa_formatter.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            anpa += (b'\x20' + sign_off) if len(sign_off) > 0 else b''
+            anpa += b'\x0D\x0A'
+
+            anpa += b'\x03'  # ETX
+
+            # time and date
+            anpa += datetime.datetime.now().strftime('%d-%m-%y %H-%M-%S').encode('ascii')
+
+            anpa += b'\x04'  # EOT
+            anpa += b'\x0D\x0A\x0D\x0A\x0D\x0A\x0D\x0A\x0D\x0A\x0D\x0A\x0D\x0A\x0D\x0A'
```
Comment (plamut) [inline server/apps/publish/formatters/anpa_formatter.py]: Repeatedly appending new chunks to an existing string is a slow (`O(n^2)`) operation. It's much better to use `bytearray`, a _mutable_ binary string data type for that ([official docs](https://docs.python.org/3/faq/programming.html#what-is-the-most-efficient-way-to-concatenate-many-strings-together)).

**Comment by plamut** [comment]
Just for the reference - I created a synthetic test and concatenating chunks using a `bytearray` is approximately 40% faster compared to using a list.

Nevertheless, both are _significantly_ faster than concatenating immutable `bytes` strings.

:+1:

**Pull request merged**